test/distributed/flight_recorder/test_fr_analysis.py

```diff
@@ -147,5 +160,83 @@ def test_match_info(self):
         self.assertEqual(str(m2), "Error type: FULLY_MATCHED, rank 1")
 
 
+LOADED_FR_DETAIL_TEMPLATE: dict[str, dict[str, Any]] = {
+    "dump_file_rank_0": {
+        "entries": [],
+        "pg_config": {
+            "0": {"name": "0", "desc": "default_pg", "ranks": "[0, 1]"},
+        },
+        "rank": 0,
+    },
+    "dump_file_rank_1": {
+        "entries": [],
+        "pg_config": {
+            "0": {"name": "0", "desc": "default_pg", "ranks": "[0, 1]"},
+        },
+        "rank": 1,
+    },
+}
+
+
+def create_one_entry(
+    record_id,
+    collective_name,
+    input_sizes,
+    output_sizes,
+    state="completed",
+    collective_seq_id=0,
+    p2p_seq_id=0,
+    output_dtypes="float32",
+):
+    event = create_one_event(
+        collective_name,
+        ("0", "default"),
+        input_sizes,
+        output_sizes,
+        state,
+        collective_seq_id,
+        p2p_seq_id,
+        output_dtypes,
+    )
+    event.update({"record_id": record_id})
+    event.update({"is_p2p": False})
+    return event
+
+
+class FlightRecorderE2ETest(TestCase):
+    def testBuildDB(self):
+        config = JobConfig()
+        args = config.parse_args([])
+        version = "2.6"  # Same as the version in FlightRecorder.hpp
+        LOADED_FR_DETAIL_TEMPLATE["dump_file_rank_0"]["version"] = version
+        LOADED_FR_DETAIL_TEMPLATE["dump_file_rank_1"]["version"] = version
+        # Test case 1: matched all_reduce case.
+        details1 = copy.deepcopy(LOADED_FR_DETAIL_TEMPLATE)
+        details1["dump_file_rank_0"]["entries"].append(
+            create_one_entry(0, "all_reduce", [[4, 4]], [[4, 4]])
+        )
+        details1["dump_file_rank_1"]["entries"].append(
+            create_one_entry(0, "all_reduce", [[4, 4]], [[4, 4]])
+        )
+        db = build_db(details1, args, version)
+        self.assertEqual(len(db.collectives), 1)
+        self.assertEqual(db.collectives[0].record_id, 0)
+        self.assertEqual(db.collectives[0].collective_name, "nccl:all_reduce")
+        self.assertEqual(db.collectives[0].pass_check, True)
+        # Test case 2: matched allreduce_coalesced case.
+        details2 = copy.deepcopy(LOADED_FR_DETAIL_TEMPLATE)
+        details2["dump_file_rank_0"]["entries"].append(
+            create_one_entry(0, "allreduce_coalesced", [[4, 4]], [[4, 4]])
+        )
+        details2["dump_file_rank_1"]["entries"].append(
+            create_one_entry(0, "allreduce_coalesced", [[4, 4]], [[4, 4]])
+        )
+        db = build_db(details2, args, version)
+        self.assertEqual(len(db.collectives), 1)
+        self.assertEqual(db.collectives[0].record_id, 0)
+        self.assertEqual(db.collectives[0].collective_name, "nccl:allreduce_coalesced")
+        self.assertEqual(db.collectives[0].pass_check, True)
```

Comment:
Do we have any integration tests running this logging with real PG NCCL and verifying that it returns expected output?

Reply:
We have written tests in test/distributed/test_c10d_nccl.py to test the FR entry content is indeed like what we listed here. And yes, ideally we want to build with real collective calls. It is to simulate different error cases is kind of tricky because there are many errors cases which are hard to enumerate. I can write a real e2e test for this as well.